### DISCUSSION TITLE
How do I update data in the production branch as part of a deploy request

### DISCUSSION THREAD

**Discussion by msevestre** [planetscale/discussion#206] [Q&A] [answered]
Hi everyone,

I have been playing with PlanetScale for the last few days and I really like the git-like branch concept. However, I haven't figured a way (yet) to run data migrations as part of the merge. 

For example, let's say that the schema migration is adding a new column to a table. And I would like this column to be filled with values computed based on other columns (as a default value just for the time being so that the updated app does not crash). In rails, I used to be able to run schema migration as part of the update as well as seeding/setting some values in the database. 

Is there a way to do something like this with PlanetScale deploy request? Or what is the suggested workflow to update schema AND data at an action? I am sure it's explained somewhere in the docs but I could not find it.

Thanks for your help,

Cheers,
Michael

**Comment by thenbe** [comment]
I've also looked through the docs but couldn't find anything that speaks to this.

In my case, I'm using prisma + planetscale and following the best practices mentioned in planetscale's [docs](https://docs.planetscale.com/tutorials/automatic-prisma-migrations) and outlined in this [video](https://youtu.be/iaHt5_hg44c). I really like the planetscale branch/deploy-request workflow to handle schema changes. But there isn't (or maybe I just couldn't find it) any documentation on how to handle the data when it comes to deploy requests.

Maybe I've missed something here, but how are other folks handling these scenarios? Take for example @msevestre's example of adding a new column and setting an initial default value from another column to avoid crashing the app. A simple example of how to handle the data in a live production db between schema changes would go a long way, for me at least. Thanks.

**Comment by itsmill3rtime** [comment]
I think because the deployment of branches is meant to be around schema versioning you aren't going to get that functionality. I use it for laravel and I have my schema migrations, then i have artisan commands i can run to perform data manipulations from cli manually after.

**Comment by iduuck** [comment]
I am also searching for a way here. I have a big migration underway, and it is strict **impossible** to do the migration without downtime, since the migration is not only **changing the schema**, but also **changing the data**. Things that could change the data is for example:

- Renaming an attribute
- Renaming a table
- "Migrating" the data from one table to another (multi-tenancy activation)
- etc.

Is there any way to solve this, otherwise (also if more expensive) I need to switch away from Planetscale :(

**Reply by mscoutermarsh** [reply]
Hi @iduuck, could you check my answer [here](https://github.com/planetscale/discussion/discussions/206#discussioncomment-4919307) and see if it helps?

In general, if your application can take downtime. Then the way to do it is. Disable the app. Run the schema migration. Run a script to migrate the data. Deploy the **new** code. Then bring the app back online.

If you're interested in ways to do this without downtime. Here's an example when renaming a column:

1. Add new column
2. Double write to old/new column
3. Backfill new column
4. Switch reads/writes 100% to new column
5. Delete the old column

Each step is a code deploy

**Comment by mscoutermarsh** [comment, marked as answer]
Hi @msevestre,

We run a Rails app ourselves. I can share how we do this. 😄 

The simple answer, is they happen separately. We run the Deploy Request to change our schema, and then run a separate rake task to complete the data migration.

Once that is complete, then we'd deploy the application code that makes use of the new column.

For anyone who is running the schema + data migration from the same script, I'd highly recommend starting to separate them out. There are a lot of benefits to this as your app grows! Having both run in a single script is fine for smaller tables. But at a certain point the data migration will take a long time, and in this time period you won't want your application code relying on this data being there (because it's highly likely it won't be yet!).

For your specific scenario. I imagine the challenge is: you need this column updated, and users are constantly writing to your app, which could change the result.

Here's how I'd accomplish this.
1. Add the new column via a deploy request (it's not used by the app yet)
2. Update my application code to write to the new column if there are any updates (still no reading from the column)
3. Run the data migration to update all missing data
4. Update my application code to make use of the new column
5. (optional) Add a null constraint to the column so that it will never be null again.

This will result in all the rows being backfilled. And you won't miss any updates that might happen between the data migration and your new app code being deployed.

I do think PlanetScale has the potential to innovate in these areas one day. The general steps above is what I'd recommend for any app though to accomplish these changes without any production downtime.

I hope this is helpful!
